### PR TITLE
fix(FTM): prevent editing in headless on preview mode

### LINKED ISSUE #30739: FTM: New Toolbar - Implement Preview Button and Preview Toolbar
### Parent Issue

https://github.com/dotCMS/core/issues/30598

### Task

Implement the Preview Button functionality and its associated Preview Toolbar mode in the new UVE Toolbar. This implementation involves creating a distinct preview mode state, managing toolbar transitions, and handling specific UI restrictions for the preview mode.

### Proposed Objective

Create a seamless preview mode experience that allows users to view their content without editing capabilities, while maintaining essential viewing tools (language and persona selectors). The preview mode should be accessible both through the Preview Button and URL parameters.

### Technical Overview

The Preview mode introduces several key changes to the UVE interface:
- Transitions the main toolbar to a simplified Preview Toolbar
- Disables editing capabilities
- Maintains language and persona selection functionality
- Introduces a new "Back to Edit Mode" button
- Handles preview mode state through both UI interaction and URL parameters

#### Preview Button Implementation
- Add Preview Button to the main toolbar following Figma designs
- Implement click handler to transition to preview mode
- Update store to track preview mode state
- Handle preview mode URL parameter (`preview=true`) on page load

#### Preview Toolbar Implementation
- Create new Preview Toolbar component that replaces main toolbar in preview mode
- Maintain Language and Persona selectors in Preview Toolbar
- Implement "Back to Edit Mode" button with proper styling
- Ensure smooth transition between regular and preview toolbars

#### Store Modifications
- Add preview mode state management to `newWithEditorToolbar()`
- Implement actions and reducers for preview mode toggle
- Create selectors to check preview mode status
- Handle persistence of preview mode state

#### Preview Mode Restrictions
- Disable and hide the following features in preview mode:
  - UVE Contentlet Palette
  - Contentlet editing tools
  - Inline editing capabilities
  - Any other editing-related functionality

#### URL Parameter Handling
- Implement logic to check for preview mode URL parameter on page load
- Update URL when entering/exiting preview mode
- Maintain other necessary URL parameters (language, persona, etc.)
- Handle proper state initialization based on URL parameters

#### Testing Requirements
- Verify preview mode activation through button click
- Test preview mode activation through URL parameters
- Ensure proper disabling of all editing features
- Test transitions between edit and preview modes
- Verify URL parameter handling and persistence
- Test across both traditional and headless pages

### Additional Considerations
- Ensure smooth animations during toolbar transitions
- Maintain proper state restoration when exiting preview mode
- Handle edge cases (e.g., invalid URL parameters)
- Consider implementing proper error handling
- Document any new store methods or state changes

### Preview Toolbar Design

In the New UVE Toolbar, this button is the trigger

Screenshot 2024-11-24 at 9 52 13 PM

And this is the toolbar to render

Screenshot 2024-11-21 at 11 06 01 AM

### Acceptance Criteria
- Clicking the trigger button should render the Preview Mode Toolbar.
- The user can access Preview Mode via a URL that includes the previewMode=true query parameter.
- While in Preview Mode, the contentlet palette and contentlet tools must be hidden, and the page should not be editable in any way.
- There should be a smooth animation when activating Preview Mode (by clicking the trigger button) and when switching between different screen resolutions.

### Resources:

[Figma component link ](https://www.figma.com/design/trbwjLuf5F6KaU1rFucNeg/Edit-Page---UVE?node-id=3488-91193&t=NzkOkm8dmgPRP0Ya-4)

### PR DESCRIPTION
**Pull request by rjvelazco**
This pull request introduces a new feature to handle preview mode across various parts of the application. The changes involve adding checks for preview mode and modifying the behavior based on whether the application is running in preview mode.

### Preview Mode Handling:

* [`core-web/libs/portlets/edit-ema/portlet/src/lib/store/features/editor/withEditor.ts`](diffhunk://#diff-86e692578757ed7f4f6cba5d0aeb07641312f3b17885825d1a45987153ae87f0L117-R120): Added a check for preview mode and modified the `isPageReady` and `isLoading` variables to account for preview mode. [[1]](diffhunk://#diff-86e692578757ed7f4f6cba5d0aeb07641312f3b17885825d1a45987153ae87f0L117-R120) [[2]](diffhunk://#diff-86e692578757ed7f4f6cba5d0aeb07641312f3b17885825d1a45987153ae87f0L134-L135)
* [`core-web/libs/sdk/client/src/lib/editor/sdk-editor.ts`](diffhunk://#diff-da7281c94a8714f497eefea7843cb58b1b25aec7513e621a0214694697ac4034R13): Introduced the `isPreviewMode` function and updated the `isInsideEditor` function to return `false` if in preview mode. [[1]](diffhunk://#diff-da7281c94a8714f497eefea7843cb58b1b25aec7513e621a0214694697ac4034R13) [[2]](diffhunk://#diff-da7281c94a8714f497eefea7843cb58b1b25aec7513e621a0214694697ac4034R113-R118)
* [`core-web/libs/sdk/client/src/lib/utils/page/common-utils.ts`](diffhunk://#diff-b7ec297f0f1414686b0318121839e205f7d0c0ad6465cfae9d41a141d0831ff9R65-R76): Added the `isPreviewMode` function to check if the application is running in preview mode based on URL parameters.
* [`dotCMS/src/main/webapp/html/js/editor-js/sdk-editor.js`](diffhunk://#diff-96d7c3f8039d9f4a1bd8a90496f8f06c3e89a4289b8d35c7d8a4941eacbfb07dL1-R1): Updated the editor initialization script to include a check for preview mode.

#### Angular

https://github.com/user-attachments/assets/ec5dd9bd-24db-4185-ab31-421009e99c85

#### ReactJS

https://github.com/user-attachments/assets/23dc8254-080a-4213-9110-d4fb38f40287

This PR fixes: #30739

### CODE REVIEW

**Review by valentinogiardino — APPROVED** [review]

File: core-web/libs/sdk/client/src/lib/editor/sdk-editor.ts [hunk 1]
```diff
@@ -109,6 +110,12 @@ export function isInsideEditor(): boolean {
         return false;
     }
 
+    const preview = isPreviewMode();
+
+    if (preview) {
+        return false;
+    }
```
Comment (zJaaal) [inline core-web/libs/sdk/client/src/lib/editor/sdk-editor.ts:117]: @fmontes This is what we were talking about the macros, should we do this and discuss it further when we work on the Arch Improvements?

Comment (zJaaal) [reply core-web/libs/sdk/client/src/lib/editor/sdk-editor.ts:117]: I also recall that we discussed about shipping preview as its own route inside edit page instead of a queryParam
 
Kevin made a POC about it https://github.com/dotCMS/core/pull/30887

Comment (rjvelazco) [reply core-web/libs/sdk/client/src/lib/editor/sdk-editor.ts:117]: Using a route instead is a good idea, but it should be well-tested. However, we will encounter a similar issue because the problem lies in the `isInsideEditor` function, which will always return `true` if the parent is not equal to the window

Comment (rjvelazco) [reply core-web/libs/sdk/client/src/lib/editor/sdk-editor.ts:117]: There is another epic to address the SDK & UVE Architecture issues. We should discuss a more complex and scalable solution for this when working on that epic.

Comment (fmontes) [reply core-web/libs/sdk/client/src/lib/editor/sdk-editor.ts:117]: There is no way that we don't have to add to the `@dotcms/client` this? for me, it looks like we're giving some kind of responsibility it should not do.

**Review by zJaaal — APPROVED** [review]

File: core-web/libs/sdk/client/src/lib/editor/sdk-editor.ts (hunk 1 above)
Comment (rjvelazco) [reply core-web/libs/sdk/client/src/lib/editor/sdk-editor.ts:117]: I'll document this!

**Review by oidacra — APPROVED** [review]
